### ISSUE TITLE
Using replace to evaluate multiple gradients during training in Keras

### ISSUE THREAD

**Issue by slavakung** [tensorflow/tensorflow#9434]
I am a researcher in optimization and I am interested in testing algorithms for training DNNs using keras, and am now using tensorflow backend.

In practice, I would like to do something a bit different from the other optimizers, I would like to compute the gradient at a slightly different value of the tensor of parameters than the current one, and the update I will make to the parameters will depend on both the current gradient and this other gradient.  

In practice this has proven more difficult than anticipated.
See https://github.com/fchollet/keras/issues/6175
it was suggested I come to here for further suggestions.

My code is a standard keras python code, the body does

model = Sequential()
model.add(Dense(512, input_shape=(784,)))
...
model.compile(loss='categorical_crossentropy',
              optimizer = myopt,
              metrics=['accuracy'])
history = model.fit(X_train, Y_train,
                    batch_size=batch_size, nb_epoch=nb_epoch, 
                    verbose=1, validation_data=(X_test, Y_test))

In the get_updates call function of my custom optimizer, it begins as usual

    def get_updates(self, params, constraints, loss):
        grads = self.get_gradients(loss, params)

Now, I want to now get the gradients at a different value of grads. First I tried just defining another tensor of the same structure but different values and take the get_gradients, but of course the loss is a graph depending on params already. Then I tried changing params itself (then copying the old values of the tensor to another one, to replace params after the evaluation) but apparently as the forward pass was not made this was ineffective. As per the advice in the above github conversation in keras, I tried,

        tempparams = [a+1. for a in params]
        replace = {p:npm for p, npm in zip(params, tempparams)}
        gradsn = [tf.contrib.graph_editor.graph_replace(g.op, replace) for g in grads]

but this is still not OK, as I get the error

TypeError: Expected a type in (<class 'tensorflow.python.framework.ops.Operation'>), got: <class 'tensorflow.python.ops.variables.Variable'

Thank you

**Comment by girving** [comment]
@purpledog: Can you glance at the last error message about `graph_replace`?  Seems like it might be an easy fix, but I'm unfamiliar with the code.

**Comment by yaroslavvb** [comment]
@slavakung rewriting the graph is a bit painful. When I do prototype non-standard optimizations I prefer keeping everything in variables. IE you run following ops 
```
gradient1.assign(current_gradient)
parameter_copy.assign(current_parameters)
current_parameters.assign(current_parameters+gradient1)
gradient3.assign(fancy_gradient(gradient1, current_gradient))
current_parameters.assign(parameter_copy)
current_parameters.assign(current_parameters+gradient3)
```
If you have these things in the loop, modifying the graph (like doing var.assign()) will add 1-10ms delay, so you may have to refactor things to pre-create those ops ahead of time.  @alextp is working on a solution that will make such applications easier

**Comment by slavakung** [comment]
I am not sure I fully understand. What is the reason I am getting the error message with graph_replace as done here? Why would it be prohibited from modifying the gradients within a get_updates call?

yaroslav,
Where in my code would/should I use assign instead?
I had heard that possibly I would have create the entire model again to begin with, which would seek computationally excessive.

**Comment by skye** [comment]
@slavakung did you manage to resolve this?

**Comment by slavakung** [comment]
No, unfortunately. I was not able to get this working in TF or Theano, and
am now working on trying to implement my algorithm in Pytorch instead. It
looks as though the tensorflow architecture is just not properly suited for
non-standard (anything but one gradient evaluation at the current point,
per iteration) optimization.


On Saturday, June 17, 2017, Skye Wanderman-Milne <notifications@github.com>
wrote:
> @slavakung did you manage to resolve this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci6.googleusercontent.com/proxy/ueJjjLeq1cdj8ageFMeNn9nU2zw6tn0o0V7-UGhiwS1jmPCwymZAyXOfnayxfpIqyv99eQURntsgepfLupGLG20mpUzafHIFKUcF8BZlCIwMXb0bsLbdd-unl4eNljiP9KXdfhnPLsZgMTzX7l7Jz-xIoY6JaA=s0-d-e1-ft#https://github.com/notifications/beacon/AEGh29DDNlCeFxf9mLn0KPOpx7QveKM_ks5sEvePgaJpZM4NHUAR.gif>

**Comment by skye** [comment]
Ah, I'm sorry to hear that. I'll leave this open for now since it hasn't been resolved, but I'm marking it "contributions welcome" since unfortunately this is a somewhat obscure use case and I don't think we have any plans to address it currently.

**Comment by namp** [comment]
Was anyone able to solve this problem?

I even copied the function 'clone_replace' from @bstriner 's [code](https://github.com/bstriner/keras-adversarial/blob/master/keras_adversarial/backend/tensorflow_backend.py) and utilized it like this:

```
grads = self.get_gradients(loss, params)
tempparams = [tf.add(a,a) for a in params]
replace = {p:npm for p, npm in zip(params, tempparams)}
gradsn = [clone_replace(g, replace) for g in grads]
```

This give the error:

`TypeError: Expected a type in (<class 'tensorflow.python.framework.ops.Operation'>), got: <class 'tensorflow.python.ops.variables.Variable'>`

If I use the `op` attribute like 

`gradsn = [clone_replace(g.op, replace) for g in grads]
`

I get the error:

`TypeError: Expected a type in (<class 'tensorflow.python.framework.ops.Tensor'>), got: <class 'tensorflow.python.framework.ops.Operation'>`

Utterly confusing...

**Comment by acutkosky** [comment]
Very late, but in case anyone else googles their way to this thread:
I came across this while looking for ways to implement SVRG in tensorflow. I think the `graph_replace` method does work for me - the only thing that needed to be fixed was the `TypeError`, because `graph_replace` takes `Tensor` as input, rather than `Variable`. I fixed this with `tf.convert_to_tensor`.
Using OP's original example, this is:
```
grads = tf.gradients(loss, params)
tempparams = [a+1. for a in params]
replace = {tf.convert_to_tensor(p) : tf.convert_to_tensor(npm) for p, npm in zip(params, tempparams)}
gradsn = tf.contrib.graph_editor.graph_replace(grads, replace)
```
